### PR TITLE
#3137 (+ #3138) - worker crash/timeout on csv export

### LINKED ISSUE #3137: [bug] OFA unable to export data to csv by record type and fiscal period
Thank you for taking the time to let us know about the issue you found. The basic rule for bug reporting is that 
something isn't working the way one would expect it to work. Please provide us with the information requested 
below and we will look at it as soon as we are able.

## Description

OFA typically extracts data for all STTs by record type and fiscal period. For some record types (e.g. T2, T3), this typically includes upwards of 500K ([reference](https://github.com/raft-tech/TANF-app/issues/2908)). on 8/8/2024, we attempted to export  the latest TANF T2 records for FY2023Q1, which includes approx 500K records, and the process failed. 

## Action Taken

- navigate to search indices table for TANF T2
- select fiscal period FY2023 Q1
- select `apply filters`
- select all records at the top of the table
- further specify intention to select all 400K+ records
- select csv export and `go` 
 
## What I expected to see

a pop-up of the csv file exported, such as in the example below:
Screenshot (25)

## What I did see

_interface_
Screenshot 2024-08-09 103922

_logs_
```shell=
10:36:18.398: [APP/PROC/WEB.0] [2024-08-09 14:36:18 +0000] [7] [WARNING] Worker with pid 1242 was terminated due to signal 9
10:36:18.405: [APP/PROC/WEB.0] [2024-08-09 14:36:18 +0000] [1419] [INFO] Booting worker with pid: 1419
```

## Other Helpful Information
+ [Demo of bug for ACF users only](https://hhsgov.sharepoint.com/:v:/r/sites/TANFDataOFA/Shared%20Documents/TANF%20Data%20System/Accessing%20TDP%20db/DAC_csv_Export_bug_demo.mp4?csf=1&web=1&e=xxWEHc) 🔏 
+ URL of the page I was on:
+ Browser and version: chrome
+ Operating System: ( Windows | MacOS X | Linux | Other ) windows
+ Is the issue repeatable?: ( yes | no | don't know ) yes
+ Has the issue occurred more than once?: yes

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #3137
* moves queryset export to a celery task and stores generated files in s3 (addressing #3138)
* implements `queryset.iterator()` to aid memory use for large querysets

## How to Test
```
cd tdrs-frontend && docker compose up --build
cd tdrs-backend && docker compose up
```

1. Open http://localhost:3000/ and sign in. Navigate to the Django Admin Console.
1. Navigate to one of the `search_indexes` model pages. select any number of models and execute the "Export as csv" action.
1. Note the returned message stating the file location. The file can be retrieved locally by navigating to `http://localhost:4566/tdp-datafiles-localstack/{returned path}`
1. Inspect the `LogEntry` objects created in the admin console.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Resolve worker crash exception preventing files from being downloaded
+ [x] Resolve worker timeout exception preventing complete queryset from being exported
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/search_indexes/tasks.py [hunk 1]
```diff
@@ -63,3 +70,92 @@ def reindex_elastic_documents():
...
+
+    try:
+        s3.client.upload_file(local_filename, settings.AWS_S3_DATAFILES_BUCKET_NAME, s3_filename)
+    except ClientError as e:
+        log(
+            f'Export failed: {s3_filename}. {e}',
+            {'user_id': system_user.pk, 'object_id': None, 'object_repr': ''},
+            'error'
+        )
+    else:
```
Comment (elipe17) [inline tdrs-backend/tdpservice/search_indexes/tasks.py:140]: This is cool, I didn't know python allowed this!

File: tdrs-backend/tdpservice/search_indexes/tasks.py [hunk 2]
```diff
@@ -63,3 +70,92 @@ def reindex_elastic_documents():
         'object_id': None,
         'object_repr': ''
     })
+
+
+class RowIterator:
```
Comment (elipe17) [inline tdrs-backend/tdpservice/search_indexes/tasks.py:74]: I was looking at this and thought we could make it a little simpler. I had a few Ideas so I stuck them into a patch for you to check out.

[export_csv.patch](https://github.com/user-attachments/files/17015206/export_csv.patch)

**Review by elipe17 — APPROVED** [review]
Works as advertised and the memory pressure is much smaller. Tested with ~400k records to export and the memory stayed very consistent. Left a couple of suggestions in a patch but they aren't a necessity. Great work, LGTM!

File: tdrs-backend/tdpservice/search_indexes/admin/mixins.py [hunk 3]
```diff
@@ -1,54 +1,13 @@
 """Mixin classes supproting custom functionality."""
-from django.contrib import admin
-from django.http import StreamingHttpResponse
-import csv
 from datetime import datetime
+from django.contrib import admin
+from ..tasks import export_queryset_to_s3_csv
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/admin/mixins.py]: please change to definite import

File: tdrs-backend/tdpservice/search_indexes/tasks.py (hunk 2 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/search_indexes/tasks.py:74]: applied! thank you

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Review by elipe17 — APPROVED** [review]
LGTM

**Comment by ADPennington** [comment]
when you get some time, can you fill in the ACs for this @jtimpe ?

**Comment by jtimpe** [comment]
> when you get some time, can you fill in the ACs for this @jtimpe ?

@ADPennington apologies - the bug tickets didn't have any AC so i added some generic ones. the implementation for each is noted in the description

**Comment by ADPennington** [comment]
> > when you get some time, can you fill in the ACs for this @jtimpe ?
> 
> @ADPennington apologies - the bug tickets didn't have any AC so i added some generic ones. the implementation for each is noted in the description

@jtimpe no trouble; mostly just trying to get a better understanding of any possible constraints when testing. will keep you posted. thank you!

**Comment by ADPennington** [comment]
first round of testing results below cc: @jtimpe :

- generated 3 exports with 3 unique s3 file paths (evidence from logentries below)
Screenshot 2024-09-18 181128

- only the most recent one comes up in the list of S3 contents:
```shell=
apennington@HHSLBDSWL73 MINGW64 /
$ aws s3 ls  s3://<bucket-name> --recursive  --profile TDP_qasp_datafilesS3_20240228 | grep -e  'exports/*'
2024-09-18 15:28:11    3035675 exports/search_indexes.ssp_m1_FY2023Q4_180924-19-25-25.csv.gz
```

I'll continue testing tmor.

**Comment by ADPennington** [comment]
2nd round of testing results cc: @jtimpe 

- retried the export of TANF_T2 records attempted yesterday (i.e. `Select all 1,091,996 TANF T2s`)
- logentries re-affirm that the export began but did not observe another entry for export completing. 
- backend logs show that worker was lost, which makes sense given the very large number of records:
```shell=
2024-09-19T08:43:38.34-0400 [APP/PROC/WEB/0] ERR [2024-09-19 12:43:38 +0000] [567] [DEBUG] Closing connection.
   2024-09-19T08:43:41.21-0400 [APP/PROC/WEB/0] ERR [2024-09-19 12:43:41,219: ERROR/MainProcess] Process 'ForkPoolWorker-10' pid:625 exited with 'signal 9 (SIGKILL)'
   2024-09-19T08:43:41.23-0400 [APP/PROC/WEB/0] ERR [2024-09-19 12:43:41,231: ERROR/MainProcess] Task handler raised error: WorkerLostError('Worker exited prematurely: signal 9 (SIGKILL) Job: 9.')
   2024-09-19T08:43:41.23-0400 [APP/PROC/WEB/0] ERR Traceback (most recent call last):
   2024-09-19T08:43:41.23-0400 [APP/PROC/WEB/0] ERR File "/home/vcap/deps/1/python/lib/python3.10/site-packages/billiard/pool.py", line 1265, in mark_as_worker_lost
   2024-09-19T08:43:41.23-0400 [APP/PROC/WEB/0] ERR raise WorkerLostError(
   2024-09-19T08:43:41.23-0400 [APP/PROC/WEB/0] ERR billiard.exceptions.WorkerLostError: Worker exited prematurely: signal 9 (SIGKILL) Job: 9.
```

will continue testing.

**Comment by jtimpe** [comment]
`top` or `htop` while ssh'd into machine during csv export (no other concurrent operations) - if mem climbs, leak somewhere - celery `DEBUG` is on

File: tdrs-backend/tdpservice/search_indexes/tasks.py [hunk 4]
```diff
@@ -63,3 +69,77 @@ def reindex_elastic_documents():
...
+    """
+    system_user, _ = User.objects.get_or_create(username='system')
+    Model = apps.get_model('search_indexes', model_name)
+    queryset = Model.objects.raw(query_str, query_params)
+    iterator = RowIterator(field_names, queryset)
+    s3 = S3Client()
+
+    tmp_filename = 'file.csv.gz'
+    record_count = -1  # offset row count to account for the header
+    with gzip.open(tmp_filename, 'wt') as f:
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/search_indexes/tasks.py:125]: reopen file on every loop iteration, append

Comment (jtimpe) [reply tdrs-backend/tdpservice/search_indexes/tasks.py:125]: ram vs disk usage

Comment (jtimpe) [reply tdrs-backend/tdpservice/search_indexes/tasks.py:125]: flush buffer

Comment (jtimpe) [reply tdrs-backend/tdpservice/search_indexes/tasks.py:125]: no effect

**Review by ADPennington — APPROVED** [review]
LGTM @jtimpe 🚀 As discussed async, detailed notes of qasp review and testing saved in [onenote](https://gorafttech.sharepoint.com/sites/TDRSResearchDesign/_layouts/15/Doc.aspx?sourcedoc={c9d41a72-bff8-4eba-b245-bb04ba265e12}&action=edit&wd=target%28Dev%20Notes.one%7C61602a1e-e3d8-4f2e-8744-c4fc299cb56b%2F3137%20qasp%20review%20resources%7C70c713b6-9724-4b9e-9c03-4aed964e476b%2F%29&wdorigin=703) 🔏 

Summary of review included below for team's visibility cc: @klinkoberstar @ttran-hub @lfrohlich 

This is a short-term solution to a critical data access-related bug; discussions are in-progress to develop long-term solution, and OFA's data team is currently relying on local db backups to access these data, so the feedback that follows can be considered for a `TBD` enhancement ticket.
- This solution works, but has some constraints:
  - the worker crashing more likely when exporting > `600K` records 
  - this solution **adds** a step into the process of accessing raw data from the db. We're moving further away from the data team's original UX with db access. 
  - the user that generates the csv export needs to keep track of the file name because it is only available to non-admin user on the search_indexes page(s). Admins can track these via LogEntries, so the user will need to rely on admins to verify if csv exports were unsuccessful (since the status is only included in LogEntries)
  - the MVP csv file naming conventions have limited value when attempting to export something other than all records by fiscal period. 
- the data team can continue relying on the local daily backup notebook for routine data access (for now), but the csv export functionality in TDP may be most useful for smaller-scale data tasks. 
- I will plan to re-test to see if any of export constraints in lower environment are present in prod.

**Pull request merged**